daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -313,21 +313,28 @@ object Hash {
     handleError(assertHashContractKey(templateId, key))
 
   // This function assumes that `arg` is well typed, i.e. :
-  // 1 - `templateId` is the identifier for a template with a contract argument of type τ
-  // 2 - `arg` is a value of type τ
+  // 1 - `packageName` is the package name defined in the metadata of the package containing template `templateId`
+  // 2 - `templateId` is the identifier for a template with a contract argument of type τ
+  // 3 - `arg` is a value of type τ
   // The hash is not stable under suffixing of contract IDs
   @throws[HashingError]
-  def assertHashContractInstance(templateId: Ref.Identifier, arg: Value): Hash =
+  def assertHashContractInstance(
+      templateId: Ref.Identifier,
+      arg: Value,
+      packageName: Ref.PackageName = Ref.PackageName.assertFromString("default"),
+  ): Hash =
     builder(Purpose.ContractInstance, aCid2Bytes)
+      .add(packageName)
```

Comment:
Why does this need to be included given that, AFAIK, the template-id contains the package-hash, which in turn also hashes the package-name?

Reply:
One requirement of the upgrade story is that some parties should be able to use (or validate transaction included) contracts that have been created with package those parties do not  know about.  In such a case this participant rely on the fact the signatories --who necessarily know about this package-- validate that the contract was created with this particular package ***and*** that the meta-data did not change during down/upgrading. Putting the meta-data in the authenticated contract ID scheme is a convenient way for the different participants to agree on it.

Reply:
Ah that's a great explanation as to why we want to include the data here as well.

I suggest adding a condensed form of that explanation as a comment to the code.